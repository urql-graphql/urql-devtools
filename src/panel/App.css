@import url("https://fonts.googleapis.com/css?family=Roboto:400,500,700&display=swap");
@import url("https://fonts.googleapis.com/css?family=Roboto+Mono:400,500,700&display=swap");

html {
  font-size: 100%;
}

body {
  font-family: "Roboto", "Helvetica", sans-serif;
  font-weight: 400;
  margin: 0;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-synthesis: none;
}

code,
pre {
  font-family: "Roboto Mono", monospace;
}

button {
  border: 0;
  padding: 0;
  background: none;
  border-radius: 0;
  font-size: inherit;
  font-family: inherit;
  cursor: pointer;
}
