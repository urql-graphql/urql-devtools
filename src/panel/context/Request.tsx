import React, {
  createContext,
  FC,
  useState,
  useEffect,
  useCallback,
  useMemo,
  useContext,
} from "react";
import {
  visit,
  buildClientSchema,
  DocumentNode,
  extendSchema,
  GraphQLSchema,
  getIntrospectionQuery,
} from "graphql";
import { gql } from "@urql/core";
import { useDevtoolsContext } from "./Devtools";

interface RequestContextValue {
  query?: string;
  setQuery: (s: string) => void;
  fetching: boolean;
  response?: Record<string, unknown>;
  execute: () => void;
  error?: Record<string, unknown>;
  schema?: GraphQLSchema;
}

export const RequestContext = createContext<RequestContextValue>(null as any);

export const useRequest = (): RequestContextValue => useContext(RequestContext);

export const RequestProvider: FC = ({ children }) => {
  const { sendMessage, addMessageHandler } = useDevtoolsContext();
  const [state, setState] = useState<{
    fetching: boolean;
    response?: Record<string, unknown>;
    error?: Record<string, unknown>;
  }>({ fetching: false, response: undefined, error: undefined });
  const [query, setQuery] = useState<string | undefined>(
    localStorage.getItem("urql-last-request") || undefined
  );
  const [schema, setSchema] = useState<GraphQLSchema>();

  const execute = useCallback(() => {
    setState({
      fetching: true,
      response: undefined,
      error: undefined,
    });
    sendMessage({
      type: "execute-query",
      source: "devtools",
      query: query || "",
    });
  }, [query, sendMessage]);

  // Listen for response for devtools
  useEffect(() => {
    return addMessageHandler((e) => {
      if (e.type !== "debug-event") {
        return;
      }

      const debugEvent = e.data;

      if (debugEvent.operation.context.meta?.source !== "Devtools") {
        return;
      }

      if (
        debugEvent.type === "update" &&
        isIntrospectionQuery(debugEvent.operation.query)
      ) {
        setSchema(
          appendPopulateDirective(buildClientSchema(debugEvent.data.value))
        );
        return;
      }

      if (debugEvent.type === "update") {
        setState({
          fetching: false,
          response: debugEvent.data.value,
        });
        return;
      }

      if (debugEvent.type === "error") {
        setState({
          fetching: false,
          error: debugEvent.data.value,
        });
        return;
      }
    });
  }, [addMessageHandler]);

  // Get schema
  useEffect(() => {
    sendMessage({
      type: "execute-query",
      source: "devtools",
      query: getIntrospectionQuery(),
    });
  }, []);

  useEffect(() => {
    if (query === undefined) {
      return;
    }
    localStorage.setItem("urql-last-request", query);
  }, [query]);

  const value = useMemo(
    () => ({
      query,
      setQuery,
      ...state,
      execute,
      schema,
    }),
    [query, state, execute, schema]
  );

  return <RequestContext.Provider value={value} children={children} />;
};

const isIntrospectionQuery = (query: DocumentNode) => {
  let value = false;

  visit(query, {
    OperationDefinition: {
      enter: (n) => {
        if (n.name?.value === "IntrospectionQuery") {
          value = true;
        }
        return false;
      },
    },
  });

  return value;
};

const appendPopulateDirective = (schema: GraphQLSchema): GraphQLSchema => {
  try {
    return extendSchema(
      schema,
      gql`
        directive @populate on FIELD
      `
    );
  } catch (err) {
    if (
      err.message.startsWith(
        'Directive "populate" already exists in the schema'
      )
    )
      return schema;
    throw err;
  }
};
