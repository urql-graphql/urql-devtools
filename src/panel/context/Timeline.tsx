import React, {
  createContext,
  Dispatch,
  useRef,
  useState,
  useCallback,
  useMemo,
  FC,
  SetStateAction,
  useContext,
  useEffect,
  useLayoutEffect,
} from "react";
import { scaleLinear, ScaleLinear } from "d3-scale";
import { DebugEvent } from "@urql/core";
import { DevtoolsContext } from "./Devtools";

interface TimelineContextValue {
  selectedEvent?: DebugEvent & { duration?: number };
  setSelectedEvent: Dispatch<SetStateAction<DebugEvent | undefined>>;
  container: HTMLDivElement;
  setContainer: (e: HTMLDivElement) => void;
  events: Record<string, DebugEvent[]>;
  eventOrder: number[];
  filterables: {
    source: string[];
    graphqlType: string[];
  };
  filter: {
    source: string[];
    graphqlType: string[];
  };
  setFilter: Dispatch<SetStateAction<TimelineContextValue["filter"]>>;
  setPosition: (time: number) => void;
  scale: ScaleLinear<number, number>;
  startTime: number;
  zoomIn: () => void;
  zoomOut: () => void;
}

export const TimelineContext = createContext<TimelineContextValue>(null as any);

export const useTimelineContext = (): TimelineContextValue =>
  useContext(TimelineContext);

const DEFAULT_WIDTH = 30000;
export const START_PADDING = 500;
const ZOOM_MIN = 5;
const ZOOM_MAX = 0.05;

const useTimelineDomain = () => {
  const startTime = useRef(Date.now());
  const domain = useRef({
    start: startTime.current - START_PADDING,
    zoom: 1,
  });
  const ref = useRef<HTMLDivElement>(undefined as any);
  const [scale, setScale] = useState<{
    scale: TimelineContextValue["scale"];
  }>({
    scale: undefined as any,
  });

  const createScale = useCallback(() => {
    if (!ref.current) {
      return;
    }

    // 30000ms at 1x zoom
    const endTime = domain.current.zoom * DEFAULT_WIDTH;

    const newScale = scaleLinear()
      .domain([domain.current.start, endTime])
      .range([0, ref.current.clientWidth]);

    setScale((oldScale) => {
      if (
        oldScale.scale &&
        // Scale hasn't changed
        newScale(100) === oldScale.scale(100) &&
        // New changes aren't visible to the user (outside of view)
        Date.now() > domain.current.start + endTime
      ) {
        return oldScale;
      }

      return {
        scale: scaleLinear()
          .domain([domain.current.start, domain.current.start + endTime])
          .range([0, ref.current.clientWidth]),
      };
    });
  }, [setScale]);

  const setPosition = useCallback((t: number) => {
    if (t < startTime.current) {
      console.warn("Cannot move position behind 'startTime'");
    }

    domain.current = {
      ...domain.current,
      start: Math.max(startTime.current - START_PADDING, t),
    };
    createScale();
  }, []);

  const setContainer = useCallback<TimelineContextValue["setContainer"]>(
    (r) => {
      ref.current = r;
      createScale();
    },
    []
  );

  const handlePan = useCallback(
    (movement: number) => {
      const endTime =
        domain.current.start + DEFAULT_WIDTH * domain.current.zoom;
      // Convert pixel delta to time delta
      const scaledDelta = scaleLinear()
        .domain([domain.current.start, endTime])
        .range([0, ref.current.clientWidth])
        .invert(movement);

      // Normalize movement for different zoom levels
      const newStart = 2 * domain.current.start - scaledDelta;

      // Apply movement (limited left movement)
      domain.current = {
        ...domain.current,
        start: Math.max(startTime.current - START_PADDING, newStart),
      };
    },
    [scale]
  );

  const handleZoom = useCallback((e: WheelEvent) => {
    e.preventDefault();
    const delta = e.deltaY * 0.01;
    const newZoom =
      e.deltaY < 0
        ? Math.max(ZOOM_MAX, domain.current.zoom + delta)
        : Math.min(ZOOM_MIN, domain.current.zoom + delta);
    const endTime = domain.current.start + domain.current.zoom * DEFAULT_WIDTH;
    const scale = scaleLinear()
      .domain([domain.current.start, endTime])
      .range([0, ref.current.clientWidth]);
    const scaleFactor = newZoom / domain.current.zoom;
    const mouseTime = scale.invert(e.clientX);
    const differenceFromStart = mouseTime - domain.current.start;
    const newDifferenceFromStart = differenceFromStart * scaleFactor;
    const newStart = mouseTime - newDifferenceFromStart;
    domain.current = {
      ...domain.current,
      start: Math.max(newStart, startTime.current - START_PADDING),
      zoom: newZoom,
    };
  }, []);

  const zoomIn = useCallback(() => {
    domain.current = {
      ...domain.current,
      zoom: Math.max(ZOOM_MAX, domain.current.zoom - 0.2),
    };
  }, []);

  const zoomOut = useCallback(() => {
    domain.current = {
      ...domain.current,
      zoom: Math.min(ZOOM_MIN, domain.current.zoom + 0.2),
    };
  }, []);

  // Listen for drag events
  useLayoutEffect(() => {
    if (!ref.current) {
      return;
    }

    // dragstart -> mousemove -> mouseup
    const dragStartListener = (e: DragEvent) => {
      e.preventDefault();
      ref.current.addEventListener("mousemove", mouseMoveListener, {
        passive: true,
      });
      window.addEventListener("mouseup", mouseUpListener, {
        passive: true,
      });
    };

    const mouseMoveListener = (e: MouseEvent) => {
      handlePan(e.movementX);
    };

    const mouseUpListener = () => {
      ref.current.removeEventListener("mousemove", mouseMoveListener);
      window.removeEventListener("mouseup", mouseUpListener);
    };

    ref.current.addEventListener("dragstart", dragStartListener);

    return () => {
      ref.current?.removeEventListener("dragstart", dragStartListener);
      ref.current?.removeEventListener("mousemove", mouseMoveListener);
      window.removeEventListener("mouseup", mouseUpListener);
    };
  }, [ref.current]);

  // Listen for zoom events
  useEffect(() => {
    if (!ref.current) {
      return;
    }

    const wheelListener = (e: WheelEvent) => {
      // Horizontal scroll
      if (!e.ctrlKey && Math.abs(e.deltaX) > Math.abs(e.deltaY)) {
        e.preventDefault();
        handlePan(-e.deltaX);
        return;
      }

      // Vertical scroll
      if (!e.ctrlKey) {
        return;
      }

      // Zoom
      e.preventDefault();
      handleZoom(e);
    };

    ref.current.addEventListener("wheel", wheelListener, { passive: false });

    return () => ref.current?.removeEventListener("wheel", wheelListener);
  }, [handleZoom, handlePan]);

  // Check scale each animation frame
  useEffect(() => {
    const render = () => {
      window.requestAnimationFrame(() => {
        createScale();
        render();
      });
    };

    render();
  }, []);

  return useMemo(
    () => ({
      container: ref.current,
      setContainer,
      startTime: startTime.current,
      setPosition,
      scale: scale.scale,
      zoomIn,
      zoomOut,
    }),
    [scale, setContainer]
  );
};

export const TimelineProvider: FC = ({ children }) => {
  const { addMessageHandler } = useContext(DevtoolsContext);
  const domain = useTimelineDomain();
  const [filterables, setFilterables] = useState<
    TimelineContextValue["filterables"]
  >({
    source: ["devtoolsExchange"],
    graphqlType: ["query", "subscription", "mutation"],
  });
  const [filter, setFilter] = useState<TimelineContextValue["filter"]>({
    source: ["devtoolsExchange"],
    graphqlType: ["query", "subscription", "mutation"],
  });
  const [events, setEvents] = useState<Record<string, DebugEvent[]>>({});
  const [eventOrder, setEventOrder] = useState<
    TimelineContextValue["eventOrder"]
  >([]);
  const [selectedEvent, setSelectedEvent] = useState<
    (DebugEvent & { duration?: number }) | undefined
  >(undefined);

  useEffect(() => {
    return addMessageHandler((message) => {
      if (message.type !== "debug-event") {
        return;
      }

      const debugEvent = message.data;
      const opKey = debugEvent.operation.key;

      setEvents((e) => ({
        ...e,
        [opKey]: [...(e[opKey] || []), message.data],
      }));
      setEventOrder((o) => (o.includes(opKey) ? o : [...o, opKey]));
      setFilterables((f) => ({
        ...f,
        source: f.source.includes(debugEvent.source)
          ? f.source
          : [...f.source, debugEvent.source],
      }));
    });
  }, [addMessageHandler]);

  const value = useMemo(
    () => ({
      events,
      eventOrder,
      setFilter,
      filter,
      filterables,
      selectedEvent,
      setSelectedEvent,
      ...domain,
    }),
    [
      domain,
      events,
      eventOrder,
      filter,
      setFilter,
      filterables,
      selectedEvent,
      setSelectedEvent,
    ]
  );

  return (
    <TimelineContext.Provider value={value}>
      {children}
    </TimelineContext.Provider>
  );
};
