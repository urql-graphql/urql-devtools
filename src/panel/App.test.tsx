jest.mock("./context/Devtools.tsx", () => {
  return {
    ...(jest.requireActual("./context/Devtools.tsx") as Record<
      string,
      unknown
    >),
    useDevtoolsContext: jest.fn(),
  };
});
import React from "react";
import { shallow } from "enzyme";
import { App, AppRoutes } from "./App";
import { useDevtoolsContext } from "./context";
import { darkTheme, lightTheme } from "./theme";

describe("App", () => {
  describe("on mount", () => {
    it("matches snapshot", () => {
      expect(shallow(<App />)).toMatchSnapshot();
    });
  });

  describe("in dark mode", () => {
    const origThemeName = chrome.devtools.panels.themeName;

    beforeAll(() => {
      chrome.devtools.panels.themeName = "dark";
    });

    afterAll(() => {
      chrome.devtools.panels.themeName = origThemeName;
    });

    it("has a GlobalStyle component and dark theme passed to ThemeProvider", () => {
      const wrapper = shallow(<App />);

      expect(wrapper.find("GlobalStyle").exists()).toBe(true);
      expect(wrapper.find("ThemeProvider").prop("theme")).toBe(darkTheme);
    });
  });

  describe("in light mode", () => {
    const origThemeName = chrome.devtools.panels.themeName;

    beforeAll(() => {
      chrome.devtools.panels.themeName = "default";
    });

    afterAll(() => {
      chrome.devtools.panels.themeName = origThemeName;
    });

    it("has a GlobalStyle component and light theme passed to ThemeProvider", () => {
      const wrapper = shallow(<App />);

      expect(wrapper.find("GlobalStyle").exists()).toBe(true);
      expect(wrapper.find("ThemeProvider").prop("theme")).toBe(lightTheme);
    });
  });
});

describe("App routes", () => {
  describe("on mount", () => {
    describe("on connected", () => {
      beforeEach(() => {
        (useDevtoolsContext as jest.Mocked<any>).mockReturnValue({
          client: {
            connected: true,
            version: {
              mismatch: false,
            },
          },
        } as any);
      });

      it("matches snapshot", () => {
        expect(shallow(<AppRoutes />)).toMatchSnapshot();
      });
    });

    describe("on version mismatch", () => {
      beforeEach(() => {
        (useDevtoolsContext as jest.Mocked<any>).mockReturnValue({
          client: {
            connected: true,
            version: {
              required: "9.9.9",
              actual: "0.0.1",
              mismatch: true,
            },
          },
        } as any);
      });

      it("matches snapshot", () => {
        expect(shallow(<AppRoutes />)).toMatchSnapshot();
      });
    });

    describe("on disconnected", () => {
      beforeEach(() => {
        (useDevtoolsContext as jest.Mocked<any>).mockReturnValue({
          client: {
            connected: false,
          },
        } as any);
      });

      it("matches snapshot", () => {
        expect(shallow(<AppRoutes />)).toMatchSnapshot();
      });
    });
  });
});
