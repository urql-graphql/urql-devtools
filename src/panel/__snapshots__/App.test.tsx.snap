// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App on mount matches snapshot 1`] = `
<ThemeProvider
  theme={
    {
      "colors": {
        "canvas": {
          "active": "#2a2a38",
          "base": "#1d1d1d",
          "elevated05": "#303030",
          "elevated10": "#373737",
          "hover": "#26252f",
        },
        "codeblock": {
          "background": "#252525",
        },
        "divider": {
          "base": "#434343",
        },
        "error": {
          "base": "#F65151",
        },
        "pending": {
          "base": "#00A1FF",
        },
        "primary": {
          "active": "#9d9cdf",
          "base": "#7776D2",
          "contrast": "#fff",
          "hover": "#8a89d8",
        },
        "secondary": {
          "base": "#FFE248",
        },
        "success": {
          "base": "#2DAF7E",
        },
        "syntax": {
          "atom": "#e8e8e8",
          "attrName": "#33b3ff",
          "base": "#33b3ff",
          "boolean": "#9291db",
          "builtin": "#9291db",
          "className": "#44b9cb",
          "comment": "#6cc7a4",
          "constant": "#33b3ff",
          "description": "#a4a4a4",
          "enum": "#6cc7a4",
          "function": "#e8e8e8",
          "input": "#9291db",
          "interface": "#9291db",
          "invalid": "#F65151",
          "keyword": "#9291db",
          "meta": "#F65151",
          "null": "#a4a4a4",
          "number": "#6cc7a4",
          "operator": "#e8e8e8",
          "property": "#33b3ff",
          "punctuation": "#e8e8e8",
          "scalar": "#ed9b3d",
          "string": "#ed9b3d",
          "type": "#33b3ff",
          "union": "#44b9cb",
          "variable": "#33b3ff",
        },
        "text": {
          "base": "#e9e9e9",
        },
        "textDimmed": {
          "active": "#d0d0d0",
          "base": "#d0d0d0",
          "hover": "#fff",
        },
        "tooltip": {
          "active": "#252525",
          "background": "#151515",
        },
      },
      "fontSizes": {
        "body": {
          "l": "0.8125rem",
          "m": "0.75rem",
          "s": "0.6875rem",
          "xl": "1rem",
        },
        "display": {
          "l": "5rem",
          "m": "3.75rem",
          "s": "2.5rem",
        },
      },
      "lineHeights": {
        "body": {
          "l": "1rem",
          "m": "0.9375rem",
          "s": "0.875rem",
          "xl": "1.25rem",
        },
        "display": {
          "l": "6rem",
          "m": "4.5rem",
          "s": "3rem",
        },
      },
      "radii": {
        "m": "0.25rem",
        "round": "50%",
        "s": "0.125rem",
      },
      "space": {
        "0": 0,
        "1": "0.125rem",
        "10": "4rem",
        "2": "0.25rem",
        "3": "0.5rem",
        "4": "0.75rem",
        "5": "1rem",
        "6": "1.25rem",
        "7": "1.5rem",
        "8": "2rem",
        "9": "3rem",
      },
    }
  }
>
  <ErrorBoundary>
    <DevtoolsProvider>
      <AppRoutes />
    </DevtoolsProvider>
  </ErrorBoundary>
  <GlobalStyle />
</ThemeProvider>
`;

exports[`App routes on mount on connected matches snapshot 1`] = `
<HashRouter>
  <Navigation
    items={
      [
        {
          "label": "Explorer",
          "link": "/explorer",
        },
        {
          "label": "Events",
          "link": "/events",
        },
        {
          "label": "Request",
          "link": "/request",
        },
      ]
    }
  />
  <TimelineProvider>
    <Route
      component={[Function]}
      path="/events"
    />
  </TimelineProvider>
  <RequestProvider>
    <Route
      component={[Function]}
      path="/request"
    />
  </RequestProvider>
  <ExplorerProvider>
    <Route
      component={[Function]}
      exact={true}
      path="/explorer"
    />
  </ExplorerProvider>
  <Route
    component={[Function]}
    exact={true}
    path="/"
  />
</HashRouter>
`;

exports[`App routes on mount on disconnected matches snapshot 1`] = `<Disconnected />`;

exports[`App routes on mount on version mismatch matches snapshot 1`] = `<Mismatch />`;
